### PR TITLE
[GST-4731] Initial setup for better configuration handling.

### PR DESCRIPTION
**Pull request by esatterwhite**
issue #4731 
- install nconf
- break fixed paths into config/overrides module
- break defaults into config/defaults module
- break database connection into separate module
- rename config#load to config#read
- move url functionality into separate module
- remove circular depencency in errors/index.js
- move deprication items into overrides
- change references to config property lookups to #get / #set
- update tests to use new configuration modules
- allow ENV vars, command line args and multipl file locations to define configuration

[GST-4731] Move database connection into separate module.

The database connection doesn't belong in the configuration, particularly
when the connection needs the configuration before it can be created.
circular things == bad time

[GST-4731] moving things that are config defaults into defaults module
- replace referecest to conf.foo to config.get('foo')
- move things out of the load / init function into defaults.
   most of the values weren't being changed and were already known

[#4731] moving remaining default configuration to defaults.js

issue #4731
- set sets default database configuration
- adds GHOST_CONFIG to initial config look up paths
- make sure paths are populated on the instance when it is exported

[#4731] rename ConfigManager#load to read

issue #4731
- nconf Provider already has a load function, rename to read

[#4731] replace config property lookups in middleware

issue #4731
    \* move config property values into defaults
    \* replace config property look ups to get function calls
    \* add spam rate limit values to defaults

[#4731] remove some console.log statements

[#4731] Update config/url.js to use config#get

issue #4731
- url module needs to use the get method and not try property lookups

config also seems like the wrong place for url helpers. Removing the
url methods from the configuration module eliminates a circular dependency

[#4731] remove the url function from configuration.

issue #4731
- remove urlFor and urlForPost from configuration
- move url module from config to utils
- change references for config.urlFor|Post to url.urlFor|Post

[#4731] Linting - No functional changes

[#4731] Updating Unit Tests for nconf style config

use config.get / set over property lookup

[#4731] fix deprecated items check.

issue #4731 the function needs to check the internal state of config,
and not try property lookup

[#4731] fix front end spec to use new config

issue #4731
- config must use get / set instead of property lookup

[#4731] Check re-configration for url first.

if a url is not specified, pull the value from config store before
calculating a subdirectory

[#4731] Correct config property look up in meta_title

issue #4731
    \* config should use get / set methods not property lookup

[#4731] correct configuration in asset, mail, naviagtion, and footer

issue #4731
- change config property lookup to user #get / #set methods

[#4731] correct configuration in asset, mail, naviagtion, and footer

issue #4731
- change config property lookup to user #get / #set methods

[#4731] 100% Passing unit tests

issue #4731
- path.join is better at dealing with leading and trailing slashes between fragements
- using path.join in favor of string concatenation.

[#4731] Correcting configuration lookups in api configuration specs

issue #4731
- configuration lookups in api configuration integration tests use #get / #set
- configuration lookups in api mail integration tests use #get / #set
- configuration lookups in api themes integration tests use #get / #set

[#4731] Linting - No functional Changes

[#4731] Fixing database references in data utils/client

issue #4931
- mysql client
- pg client
- sqlite3 client

[#4731] fix remaining legacy config references in test modules

issue #4731
- url lookup in api/users
- paths lookup in server/apps/loader
- everything in update-check & module spec
- replace old database object from conf with the db connection

[#4731] fix configuration module injection in error handling spec

issue #4731
- inject config instance, not a simple object
- errors methods should use config #get / #set over property lookup

[#4731] Use bulk insert during test runs.

posts were being inserted individually, which is really slow with
SQLite and would usually timeout, where as it would finish just fine
with postgres

[#4731] Make sure to set GHOST_CONFIG in paths if set

[#4731] Removing errors module from the config module.

issue #4731 - Config module was using the errors module to gain access
to logging methods. The logging methods needed the config module creating
a circular dependency causing one of them to not load correctly.

Logging should not be in an errors module. There needs to be a module
that is not dependant on something else with in ghost. Configuration being
the one thing that most everything else needs should be that module

[GST-4731] moving things that are config defaults into defaults module
- replace referecest to conf.foo to config.get('foo')
- move things out of the load / init function into defaults.
   most of the values weren't being changed and were already known

[#4731] remove some console.log statements

[#4731] Linting - No functional changes

[#4731] fix front end spec to use new config

issue #4731
- config must use get / set instead of property lookup

[#4731] correct configuration in asset, mail, naviagtion, and footer

issue #4731
- change config property lookup to user #get / #set methods

[#4731] correct configuration in asset, mail, naviagtion, and footer

issue #4731
- change config property lookup to user #get / #set methods

[#4731] Linting - No functional Changes

[#4731] fix remaining legacy config references in test modules

issue #4731
- url lookup in api/users
- paths lookup in server/apps/loader
- everything in update-check & module spec
- replace old database object from conf with the db connection

[#4731] fix configuration module injection in error handling spec

issue #4731
- inject config instance, not a simple object
- errors methods should use config #get / #set over property lookup

[#4731] Use bulk insert during test runs.

posts were being inserted individually, which is really slow with
SQLite and would usually timeout, where as it would finish just fine
with postgres

[#4731] Removing console statements in errors on fork utils.

noissue - no longer needed

[#4731] linting - no functional changes.

no functional changes

[#4731, #4853] including clientAssets in default configuration

issues #4853, #4731
build scripts path was removed and client assets was used in its place

[#4731] - Documentation for config modules

no functional changes

[#4731] linting - no fnctional changes

[#4731] Fix config spec to work with new config

[#4731] Correct config spec loading of defaults.

The config.example is really no longer relevant and we don't care so much.
and values can be merged. we really care about the connection specific
values in this situation

[#4731] eliminate the need to compute inside of configuration.

recalculating paths everytime something changes is a bit messy,
especially when the destination folder is always the same. Just
run path.join when you need it

[#4731] resolve config file paths.

allows for reletive or absolute locations

### CODE REVIEW

**Comment by esatterwhite** [comment]
This introduces configuration as a heirarchy. Being entirely driven by config.js, and in some cases, config.example.js

Users can set any of the configurable bits through ENV variables, command line arguments, and config files(s).

config files are not limited to what is in config.js - which now serves as the front line set of defaults. Config.js looks in a serveral locations that are common to configuration management systems like chef on unix systems for config files and applies them in an over riding fashion, as well as files based on the configured ENV, making it invariably easy to get configuration into Ghost from just about any system with any method, and A single file under source control isn't needed.

This also keep values that **should** not be changed or overridden from actually being over ridden, as overrides can not be change at all.

All of the existing tests pass, the only I wasn't able to test was the travis build. The auto generation of the config file is no longer needed. You can just update the `travis.yml` file to include the configuration in each of the entries of the matrix:

the following would configure a postgresql test run.
```
DB=pg NODE_ENV=testing database__client=pg database__connection__host=localhost database__connection__user=postgres database__connection__password='' database__connection__database=ghost_testing
```
Again, wasn't able to test this via travis, so I expect some tweating to be done.

**Comment by ErisDS** [comment]
Wow. This is an epic undertaking. I certainly think that switching towards using a standardised way of managing config is the right approach, but hopefully understandably with a PR of this size I have a couple of concerns.

My primary one is this:

> This also keep values that *should* not be changed or overridden from actually being over ridden, as overrides can not be change at all.

Could you clarify which options the PR considers to be the values that *should* not be changed? I ask this because I think we currently rely on the fact that some of the items cannot be changed in ways which are possibly wrong / will need to be changed. 

Another way to put this would be - what changes as a result of this PR? Essentially as a refactor, apart from the shiny new ways to specify config it shouldn't become possible to make fundamental changes to Ghost that weren't previously without us knowing about it.

Finally a couple of very quick comments on the code: the PR is failing on code style, which should be simple enough to fix (I think you've pushed an update whilst I'm typing this) but also your `config.example.js` changes ought not to be included, and some of the doc blocks contain @author and @require statements that don't fit with the rest of the codebase.

This is first-pass feedback, I'll take a deeper look at this later.

**Comment by esatterwhite** [comment]
Sure, anything in [overrides.js](https://github.com/esatterwhite/Ghost/blob/GST-4731/core/server/config/index.js) is considered by nconf as static and any attempts to change the value after it has initially been set, will be ignored.

So if that needs to be that way, it is a matter of adding it to overrides, or moving it from defaults, to overrides.

I also did not meat to commit `config.example.js` that was for my local testing. Will fix that 
Thanks.

**Comment by esatterwhite** [comment]
the [requires](http://usejsdoc.org/tags-requires.html) and [author](http://usejsdoc.org/tags-author.html) tags are just standard [jsdoc](http://usejsdoc.org/index.html) and make nice cross-links between pages/files. The `@author` tag is analogous to a *contributor*, there just isn't a contributor tag in jsdoc. Although, it could be changed to that without consequence. It would just be ignored.  It makes for more comprehensive and coherent code documentation. And the code base could really use that.

Are you guys using something other than jsdoc? What is it? I'd be happy to make a standards wiki doc on code documentation for the project.

**Comment by ErisDS** [comment]
@esatterwhite We're using jsdoc, loosely, but jsdoc does not require that you use all of it. We do not use the `@requires` or `@author` tags. As @sebgie has explained there is good reason not to use `@author` tags. If your PR is merged, your contribution will be recognised in the release post and on the `about` page of every single Ghost installation :wink:. Please remove the extraneous jsdoc sections from your PR.

File: package.json [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (novaugust) [inline package.json]: Sorry to be a pedant, but can this extra line die while you're modifying things? ;)

**Comment by esatterwhite** [comment]
Can any one help identify what is going on with this [build](https://travis-ci.org/TryGhost/Ghost/jobs/54934824). There seems to be a problem with something in the UI, which I didn't touch, so i'm not too sure what is wrong / what to look for

**Comment by esatterwhite** [comment]
@ErisDS, Really not concerned with recognition - I just saw an opportunity to make the documentation better.

**Comment by esatterwhite** [comment]
I updated the pull request to current master. Some of the UI tests seem to time out some of the time. If someone can re-run the tests. It is most likely fine as they pass for most of the other builds.

Also is there something else that needs to be done to merge this? I would rather not maintain this fork for much longer. It is a rather large change set.

**Comment by ErisDS** [comment]
This has been sat for quite a while now - not sure if I should just close it? I'm sorry for not having given some feedback. The large surface area of this PR made it incredibly hard to review :(

**Comment by esatterwhite** [comment]
There is a lot of surface area, but it isn't particularly complex. Basically replaced the old config thing with an nconf provider and added multiple loaders / stores so one can configure an instance anyway they want rather than the hard coded config.js in the working directory.

It is also less of a dynamic there where the old one was trying to compute paths and settings based on other paths and settings which made things really messy. The only time that really seems necessary is if you are trying to reconfigure an instance as it is running. Which seems dangerous in it self.

This set up lets you do `ghost( conf.get() )`, and not worry about file locations, names, defaults, NODE_ENV, etc. 

So one could, for example,  use chef or salt to deploy environments, and deploy an application with ghost as a dependency,
or deploy a ghost instance in a docker container and not have to worry about mounting volumes / shipping hard coded configs.
or push to heroku and use environment vars to set it up. 
or `npm install -g myapp_with_ghost` and be done.

Rebasing & Merging it would most likely be a nightmare. But the code for the config module is pretty simple. Might be easier for you guys to re-implement what is in here.

Either way, after digging around and trying to use ghost, the configuration thing seems to be at the root of many of problems that make using ghost as apart of another application ( as a module ) really difficult.

Thanks.

**Comment by ErisDS** [comment]
> Either way, after digging around and trying to use ghost, the configuration thing seems to be at the root of many of problems that make using ghost as apart of another application ( as a module ) really difficult.

I don't disagree with this statement, and I also quite like the idea of using nconf. The difficulty with this PR for me was that it proposed so many different changes it was almost impossible to fully comprehend the outcome. I don't have massive amounts of time to dedicate to non-roadmap things, as much as I wish I did because I very much appreciate the time and effort others put into trying to improve Ghost.

I think we should use this PR as a basis for how to solve #4731, and break it down into separate tasks to try to move towards this solution in a way that makes it easier to keep on top of the fast-moving codebase & to fully comprehend the consequences of each change.

@esatterwhite thank you for your work here.

**Comment by ErisDS** [comment]
FWIW - I'm still using this PR as reference material for various future-of-ghost things. There's a lot of great stuff in here, it was too dense to digest at the time, but the work hasn't gone to waste completely - I'm very much in favour of using nconf.

One thing I really do like from here is the split out of the DB connection - I'd love to see a PR that did just that piece on its own.

There seems to be 2 main commits which were involved:
- https://github.com/esatterwhite/Ghost/commit/3276163de0739de32e7d7aaba1a188193a8295eb
- https://github.com/esatterwhite/Ghost/commit/f23508717083850541c163dd8596e921745425e3

I'm mostly writing this here as a note to self :grin:

**Pull request closed without merging**